### ISSUE TITLE
`(class{f(a=1/2){}}.b=2)` is syntax error on Safari <= 16.2

### ISSUE THREAD

**Issue by hyrious** [evanw/esbuild#3072]
[see it live](https://hyrious.me/esbuild-repl/?version=0.17.17&mode=transform&input=a%3D1%2Cclass%7Bf%28b%3D1%2F2%29%7B%7D%7D.c%3D2)

```js
// input
a=1,(class{f(b=1/2){}}).c=2

// output
a = 1, class {
  f(b = 1 / 2) {
  }
}.c = 2;
```

```
// Throws error before (<=) Safari 16.2
SyntaxError: Left hand side of operator '=' must be a reference
```

**Comment by hyrious** [comment]
For now, it's ok to leave this issue to "won't fix" since Safari 16.3+ ([iOS 16.3+](https://developer.apple.com/documentation/safari-release-notes/safari-16_3-release-notes), which is [iPhone 8+](https://iosref.com/ios) since 2017) already fixed that and this is not a general JavaScript feature.

Others may change their build options to disable `--minify-syntax` to make that work on old Safari.

Feel free to close this issue.

**Comment by evanw** [comment]
Thank you for the report. I looked into this and there are a bunch of these. Some examples (I'm testing Safari 12.1 since that's the only old version I have access to):

1. `x(class{f(a=-1){}}.b=2)` is a syntax error in Safari but `x((class{f(a=-1){}}).b=2)` is not (nor is `x(class{f(a=1){}}.b=2)` so the `-` is part of the trigger)
2. `import(x+y)[0] = 0` is a syntax error in Safari but `(import(x+y))[0] = 0` is not (nor is `import(x)[0] = 0` so the `+` is part of the trigger)
3. `[+x,][0]=0` is a syntax error in Safari but `([+x,])[0]=0` is not (or is `[+x][0]=0` so the `,` is part of the trigger)

It looks like Terser with `safari10: true` doesn't handle any of these. UglifyJS with `webkit: true` handles the first one by wrapping the class in parentheses (even when it's unnecessary because Safari would parse it fine). But UglifyJS doesn't handle the other ones.

The specific workaround from UglifyJS is here: https://github.com/mishoo/UglifyJS/pull/2056. So the (overly conservative) rule is something like "wrap a function or class expression if the parent is a property access." I can do that in esbuild. And it looks like I can ignore other cases where this happens because they aren't handled by other minifiers in the ecosystem either.
